pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/TransformPipeline.java

```diff
@@ -0,0 +1,146 @@
+ */
+public class TransformPipeline {
+  private final RecordTransformer _recordTransformer;
+  private final ComplexTypeTransformer _complexTypeTransformer;
+
+  /**
+   * Constructing a transform pipeline based on TableConfig and table schema.
+   * @param tableConfig the config for the table
+   * @param schema the table schema
+   */
+  public TransformPipeline(TableConfig tableConfig, Schema schema) {
+    // Create record transformer
+    _recordTransformer = CompositeTransformer.getDefaultTransformer(tableConfig, schema);
+
+    // Create complex type transformer
+    _complexTypeTransformer = ComplexTypeTransformer.getComplexTypeTransformer(tableConfig);
+  }
+
+  /**
+   * Process and validate the decoded row against schema.
+   * @param decodedRow the row data to pass in
+   * @param reusedResult the reused result so we can reduce objects created for each row
+   * @return both processed rows and failed rows in a struct.
+   * @throws TransformException when data has issues like schema validation. Fetch the partialResult from Exception
+   */
+  public Result processRow(GenericRow decodedRow, Result reusedResult) throws TransformException {
+    // to keep track and add to "failedRows" when exception happens
+    if (reusedResult == null) {
+      reusedResult = new Result();
+    }
+    GenericRow transformedRow = null;
+    try {
+      if (_complexTypeTransformer != null) {
+        // TODO: consolidate complex type transformer into composite type transformer
+        decodedRow = _complexTypeTransformer.transform(decodedRow);
+      }
+      Collection<GenericRow> rows = (Collection<GenericRow>) decodedRow.getValue(GenericRow.MULTIPLE_RECORDS_KEY);
+      if (rows != null) {
+        for (GenericRow row : rows) {
+          transformedRow = _recordTransformer.transform(row);
+          if (transformedRow != null && IngestionUtils.shouldIngestRow(row)) {
+            reusedResult.addTransformedRows(transformedRow);
+          } else {
+            reusedResult.addFailedRows(row);
+          }
+        }
+      } else {
+        transformedRow = _recordTransformer.transform(decodedRow);
+        if (transformedRow != null && IngestionUtils.shouldIngestRow(transformedRow)) {
+          reusedResult.addTransformedRows(transformedRow);
+        } else {
+          reusedResult.addFailedRows(transformedRow);
+        }
+      }
+      return reusedResult;
+    } catch (Exception ex) {
+      // when exception happens, the current processing row needs to be added to failed list
+      reusedResult.addFailedRows(transformedRow);
+      throw new TransformException("Encountered error while processing row", reusedResult, ex);
+    }
+  }
+
+  /**
+   * Wrapper for transforming results. For efficiency, right now the failed rows have only a counter
+   */
+  public static class Result {
+    private final List<GenericRow> _transformedRows = new ArrayList<>();
+    private int _failedRowCount = 0;
```

Comment:
This is not really failed row, but dropped row (some rows are skipped intentionally but not failed)

Reply:
Renamed now